Microsoft.Toolkit.Uwp.UI/Extensions/ListViewBase/ListViewExtensions.SmoothScrollIntoView.cs

```diff
@@ -198,10 +185,52 @@ public static async Task SmoothScrollIntoViewWithIndexAsync(this ListViewBase li
         /// <param name="additionalHorizontalOffset">Adds additional horizontal offset</param>
         /// <param name="additionalVerticalOffset">Adds additional vertical offset</param>
-        /// <returns>Note: Even though this return <see cref="Task"/>, it will not wait until the scrolling completes</returns>
+        /// <returns>Returns <see cref="Task"/> that completes after scrolling</returns>
         public static async Task SmoothScrollIntoViewWithItemAsync(this ListViewBase listViewBase, object item, ScrollItemPlacement itemPlacement = ScrollItemPlacement.Default, bool disableAnimation = false, bool scrollIfVisibile = true, int additionalHorizontalOffset = 0, int additionalVerticalOffset = 0)
         {
-            await SmoothScrollIntoViewWithIndexAsync(listViewBase, listViewBase.Items.IndexOf(item), itemPlacement, disableAnimation, scrollIfVisibile, additionalHorizontalOffset, additionalVerticalOffset);
+            await SmoothScrollIntoViewWithIndexAsync(listViewBase, listViewBase.Items.IndexOf(item), itemPlacement, disableAnimation, scrollIfVisibile, additionalHorizontalOffset, additionalVerticalOffset).ConfigureAwait(true);
+        }
+
+        /// <summary>
+        /// Changes the view of <see cref="ScrollViewer"/> asynchronous.
+        /// </summary>
+        /// <param name="scrollViewer">The scroll viewer.</param>
+        /// <param name="horizontalOffset">The horizontal offset.</param>
+        /// <param name="verticalOffset">The vertical offset.</param>
+        /// <param name="zoomFactor">The zoom factor.</param>
+        /// <param name="disableAnimation">if set to <c>true</c> disable animation.</param>
+        private static async Task ChangeViewAsync(this ScrollViewer scrollViewer, double? horizontalOffset, double? verticalOffset, float? zoomFactor, bool disableAnimation)
+        {
+            var tcs = new TaskCompletionSource<VoidResult>();
+
+            void ViewChanged(object _, ScrollViewerViewChangedEventArgs e)
+            {
+                if (e.IsIntermediate)
+                {
+                    return;
+                }
+
+                tcs.TrySetResult(result: default);
+            }
+
+            try
+            {
+                scrollViewer.ViewChanged += ViewChanged;
+                scrollViewer.ChangeView(horizontalOffset, verticalOffset, zoomFactor, disableAnimation);
+                await tcs.Task.ConfigureAwait(true);
+            }
+            finally
+            {
+                scrollViewer.ViewChanged -= ViewChanged;
+            }
+        }
+
+        /// <summary>
+        /// Used as a placeholder TResult to indicate that a <![CDATA[Task<TResult>]]>  has a void TResult
+        /// </summary>
+        /// <see href="https://referencesource.microsoft.com/#System.Core/System/Threading/Tasks/TaskExtensions.cs,6e36a68760fb02e6,references"/>
+        private struct VoidResult
+        {
         }
```

Comment:
This type is unnecessary here, and it's also causing a non-shared generic instantiation of `TaskCompletionSource<T>` to be created. The previous usage of `TaskCompletionSource<object>` was completely fine and I don't really see a reason to change it. Besides, in the .NET 5 branch for the WinUI 3 target that'll just be moved to `TaskCompletionSource` (non generic) as well.

Reply:
> This type is unnecessary here

Agree. This will just prevent 4 bytes allocation in heap, that's all. I know it will be negotiable. I will remove it.

Reply:
Not sure I'm following, I don't really see how this change is avoiding any heap allocations.
Also where is that 4 bytes number coming from? Even on a 32 bit system, every allocated object would always have a size of at least 8 bytes due to the space for the object header and method table pointer. Can you elaborate? 🙂

Reply:
Sorry about the memory allocations, got confused with the generic implementation.

The below comment is from docs.microsoft.com

> The TaskCompletionSource<TResult> class doesn't have a non-generic counterpart........ (Boolean is a good default choice, but if you're concerned about the user of the Task downcasting it to a Task<TResult>, **you can use a private TResult type instead**).

Source: https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/implementing-the-task-based-asynchronous-pattern#io-bound-tasks

I understand this is overkill 🤣, just need to point this out.

Reply:
That but you quoted from the docs is outdated, there is a non-generic `TaskCompletionSource` type from .NET 5. As for your other point, I don't think we should be concerned about consumers potentially down casting to that generic task type and doing...? I don't think that'd be an issue either way, and besides we'll move to the non-generic type in the WinUI branch anyway (which has .NET 5), so this is all just temporary 🙂

Reply:
I know its outdated, I have [raised an issue in their repo](https://github.com/dotnet/docs/issues/25519)  to fix the doc issue 😉